src/Microsoft.ML.Tokenizers/PreTokenizer/TikTokenPreTokenizer.cs

```diff
@@ -0,0 +1,176 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Text.RegularExpressions;
+
+namespace Microsoft.ML.Tokenizers
+{
+    /// <summary>
+    /// The pre-tokenizer for Tiktoken tokenizer.
+    /// </summary>
+    public sealed class TikTokenPreTokenizer : PreTokenizer
+    {
+        private readonly Regex? _specialTokensRegex;
+        private readonly Regex _regex;
+
+        public TikTokenPreTokenizer(string regexPattern, IReadOnlyDictionary<string, int>? specialTokensEncoder)
+        {
+            if (regexPattern is null)
+            {
+                throw new ArgumentNullException(nameof(regexPattern));
+            }
+
+            _regex = new Regex(regexPattern, RegexOptions.Compiled);
+
+            if (specialTokensEncoder is not null && specialTokensEncoder.Count > 0)
+            {
+                _specialTokensRegex = new Regex(string.Join("|", specialTokensEncoder.Keys.Select(s => Regex.Escape(s))), RegexOptions.Compiled);
+            }
+        }
+
+        /// <summary>
+        /// Splits the given string in multiple substrings at the word boundary, keeping track of the offsets of said substrings from the original string.
+        /// </summary>
+        /// <param name="sentence">The string to split into tokens.</param>
+        /// <param name="skipSpecialTokens">Indicates whether to skip the special tokens.</param>
+        /// <returns>The list of the splits containing the tokens and the token's offsets to the original string.</returns>
+        public override IEnumerable<Split> PreTokenize(string sentence, bool skipSpecialTokens = false)
+        {
+            if (string.IsNullOrEmpty(sentence))
+            {
+                return EmptyList;
+            }
+
+            return new TokenizationEnumerable(sentence, _regex, skipSpecialTokens ? null : _specialTokensRegex);
+        }
+
+        private readonly struct TokenizationEnumerable : IEnumerable<Split>
+        {
+            private readonly string _sentence;
+            private readonly Regex _regex;
+            private readonly Regex? _specialTokensRegex;
+
+            public TokenizationEnumerable(string sentence, Regex regex, Regex? specialTokensRegex)
```

Comment:
Parsing via regex is _the_ way to do this? There's no dedicated, specialized parsing routine that should be used instead?

Reply:
If need to have a dedicated parsing routine this will be same as re-implementing regex parsing. The current LLP models is using regex patterns for their tokenization process. For example GPT-4 is using 
`@"(?i:'s|'t|'re|'ve|'m|'ll|'d)|[^\r\n\p{L}\p{N}]?\p{L}+|\p{N}{1,3}| ?[^\s\p{L}\p{N}]+[\r\n]*|\s*[\r\n]+|\s+(?!\S)|\s+";`. I don't think we need to go to that level providing out own parsing.